README.md
# InfluxDB Exporter [![Build Status](https://travis-ci.org/prometheus/influxdb_exporter.svg)][travis]

[![CircleCI](https://circleci.com/gh/prometheus/influxdb_exporter/tree/master.svg?style=shield)][circleci]
[![Docker Repository on Quay](https://quay.io/repository/prometheus/influxdb-exporter/status)][quay]
[![Docker Pulls](https://img.shields.io/docker/pulls/prom/influxdb-exporter.svg?maxAge=604800)][hub]

An exporter for metrics in the InfluxDB format used since 0.9.0. It collects
metrics in the
[line protocol][line_protocol] via a HTTP API,
transforms them and exposes them for consumption by Prometheus.

This exporter supports float, int and boolean fields. Tags are converted to Prometheus labels.

The exporter also listens on a UDP socket, port 9122 by default.

## Timestamps

By default metrics exposed without original timestamps like this:

```
http_requests_total{method="post",code="200"} 1027
http_requests_total{method="post",code="400"}    3
```

If you want to add original timestamps to exposed metrics, please use flag `--timestamps` and metrics will looks like:

```
http_requests_total{method="post",code="200"} 1027 1395066363000
http_requests_total{method="post",code="400"}    3 1395066363000
```

When querying, this means that the sample is attributed to the time it was
submitted to the exporter, not the time Prometheus scraped it. However, if the
metric was submitted multiple time in between exporter scrapes, only the last
value and timestamp will be stored.

## Alternatives

If you are sending data to InfluxDB in Graphite or Collectd formats, see the
[graphite_exporter][graphite_exporter]
and [collectd_exporter][collectd_exporter] respectively.

This exporter is useful for exporting metrics from existing collectd setups, as
well as for metrics which are not covered by the core Prometheus exporters such
as the [Node Exporter][node_exporter].

The exporter acts like an InfluxDB server, it does not connect to one. For
metrics concerning the InfluxDB server, use the [metrics endpoint][influxdb_metrics]
built into InfluxDB.

If you are already using Telegraf, it can serve the same purpose as this
exporter with the [`outputs.prometheus_client`][telegraf] plugin.

For more information on integrating between the Prometheus and InfluxDB
ecosystems, see the [influxdata integration page][influx_integration].

## Example usage with Telegraf

The influxdb_exporter appears as a normal InfluxDB server. To use with Telegraf
for example, put the following in your `telegraf.conf`:

```
[[outputs.influxdb]]
  urls = ["http://localhost:9122"]
```

Or if you want to use UDP instead:
```
[[outputs.influxdb]]
  urls = ["udp://localhost:9122"]
```

Note that Telegraf already supports outputing Prometheus metrics over HTTP via
[`outputs.prometheus_client`][telegraf], which avoids having to also run the influxdb_exporter.


[circleci]: https://circleci.com/gh/prometheus/influxdb_exporter
[hub]: https://hub.docker.com/r/prom/influxdb-exporter/
[travis]: https://travis-ci.org/prometheus/influxdb_exporter
[quay]: https://quay.io/repository/prometheus/influxdb-exporter
[line_protocol]: https://docs.influxdata.com/influxdb/v0.10/write_protocols/line/
[graphite_exporter]: https://github.com/prometheus/graphite_exporter
[collectd_exporter]: https://github.com/prometheus/collectd_exporter
[node_exporter]: https://github.com/prometheus/node_exporter
[influxdb_metrics]: https://docs.influxdata.com/influxdb/v1.5/administration/server_monitoring/#influxdb-metrics-http-endpoint
[telegraf]: https://docs.influxdata.com/telegraf/v1.7/plugins/outputs/#prometheus-client-prometheus-client-https-github-com-influxdata-telegraf-tree-release-1-7-plugins-outputs-prometheus-client
[influx_integration]: https://www.influxdata.com/integration/prometheus-monitoring-tool/


CHANGELOG.md
## ?.?.? / unreleased

* [CHANGE] Do not run as root in the Docker container by default ([#40](https://github.com/prometheus/influxdb_exporter/pull/40))

## 0.2.0 / 2019-02-28

* [CHANGE] Switch to Kingpin flag library ([#14](https://github.com/prometheus/influxdb_exporter/pull/14))
* [FEATURE] Optionally export samples with timestamp ([#36](https://github.com/prometheus/influxdb_exporter/pull/36))

For consistency with other Prometheus projects, the exporter now expects
POSIX-ish flag semantics. Use single dashes for short options (`-h`) and two
dashes for long options (`--help`).

## 0.1.0 / 2017-07-26

Initial release.


main.go
// Copyright 2016 The Prometheus Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"regexp"
	"sort"
	"sync"
	"time"

	"gopkg.in/alecthomas/kingpin.v2"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"

	"github.com/influxdata/influxdb/models"
)

const (
	MAX_UDP_PAYLOAD = 64 * 1024
)

var (
	listenAddress   = kingpin.Flag("web.listen-address", "Address on which to expose metrics and web interface.").Default(":9122").String()
	metricsPath     = kingpin.Flag("web.telemetry-path", "Path under which to expose Prometheus metrics.").Default("/metrics").String()
	sampleExpiry    = kingpin.Flag("influxdb.sample-expiry", "How long a sample is valid for.").Default("5m").Duration()
	bindAddress     = kingpin.Flag("udp.bind-address", "Address on which to listen for udp packets.").Default(":9122").String()
	exportTimestamp = kingpin.Flag("timestamps", "Export timestamps of points").Default("false").Bool()
	lastPush        = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "influxdb_last_push_timestamp_seconds",
			Help: "Unix timestamp of the last received influxdb metrics push in seconds.",
		},
	)
	udpParseErrors = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "influxdb_udp_parse_errors_total",
			Help: "Current total udp parse errors.",
		},
	)
	invalidChars = regexp.MustCompile("[^a-zA-Z0-9_]")
)

type influxDBSample struct {
	ID        string
	Name      string
	Labels    map[string]string
	Value     float64
	Timestamp time.Time
}

func (c *influxDBCollector) serveUdp() {
	buf := make([]byte, MAX_UDP_PAYLOAD)
	for {
		n, _, err := c.conn.ReadFromUDP(buf)
		if err != nil {
			log.Warnf("Failed to read UDP message: %s", err)
			continue
		}

		bufCopy := make([]byte, n)
		copy(bufCopy, buf[:n])

		precision := "ns"
		points, err := models.ParsePointsWithPrecision(bufCopy, time.Now().UTC(), precision)
		if err != nil {
			log.Errorf("Error parsing udp packet: %s", err)
			udpParseErrors.Inc()
			return
		}

		c.parsePointsToSample(points)
	}
}

type influxDBCollector struct {
	samples map[string]*influxDBSample
	mu      sync.Mutex
	ch      chan *influxDBSample

	// Udp
	conn *net.UDPConn
}

func newInfluxDBCollector() *influxDBCollector {
	c := &influxDBCollector{
		ch:      make(chan *influxDBSample),
		samples: map[string]*influxDBSample{},
	}
	go c.processSamples()
	return c
}

func (c *influxDBCollector) influxDBPost(w http.ResponseWriter, r *http.Request) {
	lastPush.Set(float64(time.Now().UnixNano()) / 1e9)
	buf, err := ioutil.ReadAll(r.Body)
	if err != nil {
		http.Error(w, fmt.Sprintf("error reading body: %s", err), 500)
		return
	}

	precision := "ns"
	if r.FormValue("precision") != "" {
		precision = r.FormValue("precision")
	}
	points, err := models.ParsePointsWithPrecision(buf, time.Now().UTC(), precision)
	if err != nil {
		http.Error(w, fmt.Sprintf("error parsing request: %s", err), 400)
		return
	}

	c.parsePointsToSample(points)

	// InfluxDB returns a 204 on success.
	http.Error(w, "", http.StatusNoContent)
}

func (c *influxDBCollector) parsePointsToSample(points []models.Point) {
	for _, s := range points {
		fields, err := s.Fields()
		if err != nil {
			log.Errorf("error getting fields from point: %s", err)
			continue
		}
		for field, v := range fields {
			var value float64
			switch v := v.(type) {
			case float64:
				value = v
			case int64:
				value = float64(v)
			case bool:
				if v {
					value = 1
				} else {
					value = 0
				}
			default:
				continue
			}

			var name string
			if field == "value" {
				name = string(s.Name())
			} else {
				name = fmt.Sprintf("%s_%s", s.Name(), field)
			}

			sample := &influxDBSample{
				Name:      invalidChars.ReplaceAllString(name, "_"),
				Timestamp: s.Time(),
				Value:     value,
				Labels:    map[string]string{},
			}
			for _, v := range s.Tags() {
				sample.Labels[invalidChars.ReplaceAllString(string(v.Key), "_")] = string(v.Value)
			}

			// Calculate a consistent unique ID for the sample.
			labelnames := make([]string, 0, len(sample.Labels))
			for k := range sample.Labels {
				labelnames = append(labelnames, k)
			}
			sort.Strings(labelnames)
			parts := make([]string, 0, len(sample.Labels)*2+1)
			parts = append(parts, name)
			for _, l := range labelnames {
				parts = append(parts, l, sample.Labels[l])
			}
			sample.ID = fmt.Sprintf("%q", parts)

			c.ch <- sample
		}
	}
}

func (c *influxDBCollector) processSamples() {
	ticker := time.NewTicker(time.Minute).C
	for {
		select {
		case s := <-c.ch:
			c.mu.Lock()
			c.samples[s.ID] = s
			c.mu.Unlock()

		case <-ticker:
			// Garbage collect expired value lists.
			ageLimit := time.Now().Add(-*sampleExpiry)
			c.mu.Lock()
			for k, sample := range c.samples {
				if ageLimit.After(sample.Timestamp) {
					delete(c.samples, k)
				}
			}
			c.mu.Unlock()
		}
	}
}

// Collect implements prometheus.Collector.
func (c *influxDBCollector) Collect(ch chan<- prometheus.Metric) {
	ch <- lastPush

	c.mu.Lock()
	samples := make([]*influxDBSample, 0, len(c.samples))
	for _, sample := range c.samples {
		samples = append(samples, sample)
	}
	c.mu.Unlock()

	ageLimit := time.Now().Add(-*sampleExpiry)
	for _, sample := range samples {
		if ageLimit.After(sample.Timestamp) {
			continue
		}

		metric := prometheus.MustNewConstMetric(
			prometheus.NewDesc(sample.Name, "InfluxDB Metric", []string{}, sample.Labels),
			prometheus.UntypedValue,
			sample.Value,
		)

		if *exportTimestamp {
			metric = prometheus.NewMetricWithTimestamp(sample.Timestamp, metric)
		}
		ch <- metric
	}
}

// Describe implements prometheus.Collector.
func (c *influxDBCollector) Describe(ch chan<- *prometheus.Desc) {
	ch <- lastPush.Desc()
}

func init() {
	prometheus.MustRegister(version.NewCollector("influxdb_exporter"))
	prometheus.MustRegister(udpParseErrors)
}

func main() {
	log.AddFlags(kingpin.CommandLine)
	kingpin.Version(version.Print("influxdb_exporter"))
	kingpin.HelpFlag.Short('h')
	kingpin.Parse()

	log.Infoln("Starting influxdb_exporter", version.Info())
	log.Infoln("Build context", version.BuildContext())

	c := newInfluxDBCollector()
	prometheus.MustRegister(c)

	addr, err := net.ResolveUDPAddr("udp", *bindAddress)
	if err != nil {
		fmt.Printf("Failed to resolve UDP address %s: %s", *bindAddress, err)
		os.Exit(1)
	}

	conn, err := net.ListenUDP("udp", addr)
	if err != nil {
		fmt.Printf("Failed to set up UDP listener at address %s: %s", addr, err)
		os.Exit(1)
	}

	c.conn = conn
	go c.serveUdp()

	http.HandleFunc("/write", c.influxDBPost)
	// Some InfluxDB clients try to create a database.
	http.HandleFunc("/query", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, `{"results": []}`)
	})

	http.Handle(*metricsPath, promhttp.Handler())

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
    <head><title>InfluxDB Exporter</title></head>
    <body>
    <h1>InfluxDB Exporter</h1>
    <p><a href="` + *metricsPath + `">Metrics</a></p>
    </body>
    </html>`))
	})

	log.Infoln("Listening on", *listenAddress)
	http.ListenAndServe(*listenAddress, nil)
}
